### PR TITLE
Puzzle mode options

### PR DESCRIPTION
**Pull request by EvidentlyCube**
[Relevant thread](http://forum.caravelgames.com/viewtopic.php?TopicID=44932)

Note: I have not moved any texts to MIDs file yet. I'll do it tomorrow, but in the meantime feel free to review the rest of the code.

### CODE REVIEW

**Comment by EvidentlyCube** [comment]
I consider this finished.

File: .gitignore [hunk 1]
```diff
@@ -79,7 +79,11 @@ drodrpg/DRODUtil/Release-Temp/
 **/*.exe
 **/*.vcxproj.user
 **/*.vcxproj.filters
+!drod/drod.2019.vcxproj.filters
```
Comment (mrimer) [inline .gitignore:82]: Thanks for catching these.

File: DROD/GameScreen.cpp [hunk 2]
```diff
@@ -6311,6 +6329,31 @@ void CGameScreen::ShowSpeechLog()
 	Paint();
 }
 
+//*****************************************************************************
+void CGameScreen::ShowPuzzleModeOptions()
+//Pop up a dialog box that contains a log of the speech played in the current room.
```
Comment (mrimer) [inline DROD/GameScreen.cpp:6334]: Need to update comment.

File: DROD/RoomWidget.cpp [hunk 3]
```diff
@@ -2128,24 +2128,57 @@ void CRoomWidget::ShowMoveCount(const bool bVal)
...
+//*****************************************************************************
+void CRoomWidget::SetPuzzleModeOptions(const PuzzleModeOptions &puzzleModeOptions)
+//Enables puzzle mode (display for easier puzzle solving).
+{
+	this->puzzleModeOptions = puzzleModeOptions;
+	this->puzzleModeOptions.bIsEnabled = this->bShowPuzzleMode;
+
+	if (this->bShowPuzzleMode) {
+		TogglePuzzleMode();
+		TogglePuzzleMode();
```
Comment (mrimer) [inline DROD/RoomWidget.cpp:2141]: Please comment why this needs to be called twice.

File: FrontEndLib/LabelWidget.h [hunk 4]
```diff
@@ -33,6 +33,13 @@
 #include <string>
 using std::string;
 
+//Y position for a label using F_TITLE font on the screens with the darkened horizontal bar at the top
+// to fit neatly in the middle of said bar
+static const UINT Y_TITLE_LABEL_CENTER_DARK_BAR = 16;
+//Standard height of labels.
+static const UINT CY_LABEL_FONT_TITLE = 52;
+static const UINT CY_LABEL_FONT_HEADER = 30;
```
Comment (mrimer) [inline FrontEndLib/LabelWidget.h:41]: Thank you for refactoring these magic numbers.

File: drod/PuzzleModeOptions.h [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+#ifndef PUZZLEMODEOPTIONS_H
+#define PUZZLEMODEOPTIONS_H
+
+#include <BackEndLib/Types.h>
+#include <DRODLib/DbPackedVars.h>
+#include <DRODLib/SettingsKeys.h>
+
+struct PuzzleModeOptions {
+	PuzzleModeOptions() {
```
Comment (mrimer) [inline drod/PuzzleModeOptions.h:36]: For the default constructor, lets use the standard construction variable initialization idiom, i.e.,
`Constructor() : var1(val), var2(val), ... { }`

**Review by mrimer — CHANGES_REQUESTED** [review]
Puzzle mode fans now have everything their heart desires!
Thank you.  Looks great.
In noted a few minor changes to make.

**Comment by EvidentlyCube** [comment]
PR updated!

**Review by mrimer — APPROVED** [review]
Thanks for the revisions!

**Pull request merged**